client/synapseJavaClient/src/main/java/org/sagebionetworks/client/AsynchJobType.java

```diff
@@ -42,20 +45,58 @@
 	
 	/**
 	 * Get the URL used to start this job type.
+	 * @param request
 	 */
-	public  String getStartUrl(){
-		return prefix+ASYNC_START;
+	public  String getStartUrl(AsynchronousRequestBody request){
+		String entityId = getEntityIdFromRequest(request);
+		if (entityId != null) {
+			return "/entity/" + entityId + prefix + ASYNC_START;
+		} else {
+			return prefix + ASYNC_START;
+		}
+	}
+
+	/*
+	 * extracts the entityId from the request body 
+	 * throws an exception if the request has an entityId field but the entityId is null
+	 * If the request body does not have an entityId field, returns null.
+	 */
+	private String getEntityIdFromRequest(AsynchronousRequestBody request) {
+		if (request instanceof UploadToTableRequest && ((UploadToTableRequest) request).getTableId() != null) {
+			return ((UploadToTableRequest) request).getTableId();
+		} else if (request instanceof HasEntityId && ((HasEntityId) request).getEntityId() != null) {
+			return ((HasEntityId) request).getEntityId();
+		} else if ((request instanceof UploadToTableRequest && ((UploadToTableRequest) request).getTableId() == null) ||
+					(request instanceof HasEntityId && ((HasEntityId) request).getEntityId() == null)) {
+			throw new IllegalArgumentException("entityId cannot be null");
+		} else {
+			return null;

```

Comment:
At line 71, one of the following is true:
+ the request is an instance of UploadToTableRequest and its tableId field is null; or
+ the request is an instance of HasEntityId and its entityId field is null
At line 73, "else" means: the request is not an instance of UploadToTableRequest and also not an instance of HasEntityId. In this case, I do not care about getting the entityId field from this request. So the method returns null.

The way my mind works is using nested if-else. But I heard that nested if-else is bad, so I tried to write it this way. Does it make sense now? Does the comment above the method help at all? How should I write this better?